### ISSUE TITLE
Multiple Learning Rates

### ISSUE THREAD

**Issue by mkohram** [fchollet/keras#414]
What would be the simplest method to perform back propagation on the top n layers? I want to be able to stop learning at the bottom layers for a while. This can be done either by setting a small learning curve on these layers or back propagating only on the top layers. I'm new to Keras/Theano so any pointers would be appreciated.

Thanks

**Comment by fchollet** [comment]
Your best shot without modifying Keras in-depth would be weight reuse. Start by training a shallow network, then add new layers on top of it while initializing the weights of the lower layers with that of your previously trained layers.

You can use: `layer.get_weights()` and `layer.set_weights(weights)`

The alternative would be to modify the optimizer system in Keras to pass an argument that specifies each layer's depth, then apply different updates in the optimizer depending on the depth. It wouldn't be too difficult to pull off, but it would require modifying a thing or two. I can give you pointers if you want to do that.

**Comment by fchollet** [comment]
Actually, I think I misunderstood your question; freezing learning in the lower layers can't be achieved by gradually stacking up new layers... what you can do, however is record the output of a lower layer then train the upper part of the network on these outputs. Then only the upper layers are being modified.

**Comment by mkohram** [comment]
dammit why didn't I think of that! Thanks.

**Comment by fchollet** [comment]
Not sure what you mean. But here's how you can fully control learning speed by depth in any optimizer:

- in every layer you use: add a constructor argument `depth`, set it as layer attribute
- in `layers.core.Layer`, method `get_params`: also return a depths array the same as `params`
- in `layers.containers.Sequential`, method `add`: make sure a class attribute `self.depths` is being updated
- in `models.Sequential`, method `compile`: replace
```python
updates = self.optimizer.get_updates(self.params, self.constraints, train_loss)
```
with 
```python
updates = self.optimizer.get_updates(self.params, self.constraints, self.depths, train_loss)
```
- in all optimizers: take into account this new `depths` argument to `get_updates` to configure the gradient updates.

**Comment by mkohram** [comment]
This is very useful. Thanks a lot.

**Comment by Mofef** [comment]
These were two very useful tips. I'm commenting here to say thanks and also to keep this conversation. btw: Is there another way to "bookmark" such a thing in github?

**Comment by jeanpijon** [comment]
Hi, first of all - thank you for such a great framework. Second - I have working implementation of lr_scaling - layer wise. It works similarly like depths argument, but with much cleaner semantics. layer can get in initialization lr_scaler which get distributed up to optimizer. In optimizers LR for such layer is simply multiplied by such scaler. Do you think, it would be beneficial for the community? I can do a pull request, if there would be demand for it.

**Comment by MarcBS** [comment]
Hi @jeanpijon, is the feature you mentioned about lr_scaling currently implemented?
Thanks.

**Comment by jeanpijon** [comment]
I was overthinking this issue again - LR is more a function of optimizer, not Layers itself. so there should be parameter to supply into optimizer constructor. I am suggesting dictionary with semantics: {layer_num: lr_scale, ...} I do have this modification and can submit a PR.

**Comment by AvantiShri** [comment]
@MarcBS If you are interested in an implementation that does make the learning rate multiplier a function of the layer itself, my lab has it implemented on our fork of keras (the arguments are W_learning_rate_multiplier and b_learning_rate_multiplier for fully connected or convolutional layers: https://github.com/kundajelab/keras/blob/master/keras/layers/convolutional.py#L75). We frequently sync our repo to the master branch of the main keras repo.

The reason I haven't submitted a pull request is that I haven't found the time to write unit tests, and because there's a few other additions to our lab's keras repo (namely task-specific weights for sequential models, and some comp-bio specific initialisation schemes/constraints) that I'd have to isolate away.

**Comment by MarcBS** [comment]
Thank you both for your response and your proposals, I think that either having a weighted learning rate directly as a layer parameter or in an argument for the optimizer are good ideas.
Keep us updated if you submit a PR @jeanpijon .

Meanwhile I have taken a look at your fork @AvantiShri , it looks like right now you can only provide the multipliers at the layers initialization. But I suppose it should be straightforward to add a setter for a posterior modification, is it?

**Comment by AvantiShri** [comment]
@MarcBS Yeah; the critical thing to modify, in addition to self.W_learning_rate_multiplier and self.b_learning_rate_multiplier would be self.learning_rate_multipliers, since that is what is picked up by get_params() in https://github.com/kundajelab/keras/blob/master/keras/layers/core.py#L276 (I mimicked the way constraints are handled).
